### PR TITLE
Deprecate --recompile-scripts

### PR DESCRIPTION
**Pull request by blindpirate**
## Context

See #1425

### CODE REVIEW

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaFirstUsePerformanceTest.groovy [hunk 1]
```diff
@@ -30,7 +30,6 @@ class JavaFirstUsePerformanceTest extends AbstractCrossVersionPerformanceTest {
         runner.testProject = testProject
         runner.gradleOpts = ["-Xms${testProject.daemonMemory}", "-Xmx${testProject.daemonMemory}"]
         runner.tasksToRun = ['tasks']
-        runner.args = ['--recompile-scripts'] // This is an approximation of first use: we recompile the scripts
```
Comment (oehme) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaFirstUsePerformanceTest.groovy]: If we just remove this, then this test no longer makes any sense. We need to add an experiment listener that deletes the script cache after the build.

**Comment by bmuschko** [comment]
@blindpirate @blindpirate [My PR](https://github.com/gradle/gradle/pull/2890) already provides an API for deprecating command line options. We should merge that first. Then you don't have to use `DeprecationLogger` directly.

**Comment by blindpirate** [comment]
@bmuschko Thanks for your kind reminder!

**Comment by bmuschko** [comment]
@blindpirate My changes have been merged. You will want to rebase on `master`. To deprecate the command line option you just have to call `CommandLineOptionConfiguration.deprecated(String)`.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaFirstUsePerformanceTest.groovy (hunk 1 above)
Comment (blindpirate) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaFirstUsePerformanceTest.groovy]: Given that we're deprecating things instead of removing, I keep this flag unchanged.

**Comment by blindpirate** [comment]
@bmuschko Thanks for your work, it makes things much easier.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy [hunk 2]
```diff
@@ -93,6 +93,7 @@ class CacheProjectIntegrationTest extends AbstractIntegrationTest {
         classFile.assertHasChangedSince(classFileSnapshot)
         classFileSnapshot = classFile.snapshot()
 
+        executer.expectDeprecationWarning().requireGradleDistribution()
```
Comment (oehme) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy]: ❌ Why was `requireGradleDistribution` added here?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 3]
```diff
@@ -1145,6 +1145,9 @@ private void validate(String output, String displayName) {
                         while (i < lines.size() && STACK_TRACE_ELEMENT.matcher(lines.get(i)).matches()) {
                             i++;
                         }
+                    } else if (line.matches(".*\\[deprecated.*]")) {
```
Comment (oehme) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: ❌ Why was this added?

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy (hunk 2 above)
Comment (blindpirate) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy]: Without `requireGradleDistribution`, the warning will be printed at the same JVM with the test code, thus we can't catch it.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 3 above)
Comment (blindpirate) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: After @bmuschko 's #2880 , if we have a deprecated option, when we run `gradle --help`, we'll get:

```
USAGE: gradle [option...] [task...]

-?, -h, --help          Shows this help message.
...
--recompile-scripts     Force build script recompiling. [deprecated this has been deprecated]
...
```

Since this output contains `deprecated`, it would be treated as deprecation warning, which made many tests fail.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy (hunk 2 above)
Comment (oehme) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy]: I don't think that's the case, there are plenty of tests using `expectDeprecationWarning`, but not `requireGradleDistribution`.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 3 above)
Comment (oehme) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Instead of adding comments, consider using a descriptive method for such conditions like `if (isDescriptionOfDeprecatedOption())`. Also consider nesting this condition inside the next one, so their order doesn't matter.

Comment (blindpirate) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Good advice.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy (hunk 2 above)
Comment (blindpirate) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy]: Things are a bit different here. In this case, the warning message is outputted during parameter parsing, which happens before we can catch the build message: https://github.com/gradle/gradle/blob/master/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecuter.java#L267

Comment (blindpirate) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy]: In other cases, the outputted deprecation message would be catched by this: https://github.com/gradle/gradle/blob/master/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecuter.java#L279

That's why other build operations don't need isolated distribution.

Comment (oehme) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/CacheProjectIntegrationTest.groovy]: Ah, that's what I was missing. Thanks for the explanation!

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/RecompileScriptsDeprecationIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+package org.gradle.initialization
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
+
+@Issue('https://github.com/gradle/gradle/issues/1425')
+class RecompileScriptsDeprecationIntegrationTest extends AbstractIntegrationSpec {
+    def "deprecation warning appears when using --recompile-scripts"() {
+        given:
+        buildFile << 'task hello{ doLast{ println "hello" } }'
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/initialization/RecompileScriptsDeprecationIntegrationTest.groovy]: We usually use multi-line build script definitions for readability. Not critical, just something to be aware of. No need to necessarily change it here.

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptionFactory.java [hunk 5]
```diff
@@ -85,8 +85,9 @@ public void applyTo(StartParameter settings, Origin origin) {
     }
 
     public static class RecompileScriptsOption extends EnabledOnlyBooleanBuildOption<StartParameter> {
+        public static final String DEPRECATION_MESSAGE = "Support for --recompile-scripts was deprecated and is scheduled to be removed in Gradle 5.0.";
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptionFactory.java]: Wondering if we could separate the message generation logic in `DeprecationLogger` from the console logging aspect. Then you could use the message generation here without having to hard-code the message in this class.
